trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/container/c/CArray.java

```diff
@@ -71,31 +63,43 @@ public final class CArray implements ICType {
 	 * @param valueType
 	 *            the type of the array.
 	 */
-	public CArray(final RValue bound, final ICType valueType) {
+	public CArray(final Expression bound, final ICType boundType, final ICType valueType) {
 		mBound = bound;
+		mBoundType = boundType;
 		mValueType = valueType;
 	}
 
 	/**
 	 * @return the dimensions
 	 */
-	public RValue getBound() {
+	public Expression getBound() {
 		return mBound;
 	}
 
+	public ICType getBoundType() {
+		return mBoundType;
+	}
+
 	/**
 	 * @return the valueType
 	 */
 	public ICType getValueType() {
 		return mValueType;
 	}
 
+	public void complete(final Expression bound) {
+		if (mBound != null) {
+			throw new AssertionError("only incomplete arrays can be completed");
+		}
+		mBound = bound;
+	}
+

```

Comment:
Making the type mutable in this way seems risky. Types that were previously considered equal will be non-equal after a call to this method. The hashcode will change, which is dangerous if the type is stored in some hash set or hash map.